.github/workflows/deploy-docs.yml

```diff
@@ -8,26 +8,95 @@ on:
     paths:
       - '.github/workflows/deploy-docs.yml'
       - 'docs/**'
+    # NOTE: The tag push event is not constrained by branch and path `on` conditions.
+    # For taking a snapshot of the docs state from the tagged commit,
+    # to a versioned subdirectory with dir name matching the git tag substring `v<MAJOR>.<MINOR>`:
+    tags:
+      - 'v[0-9]+.[0-9]+*'
 
-# Jobs will run shell commands from this subdirectory:
-defaults:
-  run:
-    working-directory: docs
+env:
+  # Default docs version to build and deploy
+  DOCS_VERSION: edge
+  GIT_USER: 'github-actions[bot]'
+  GIT_EMAIL: '41898282+github-actions[bot]@users.noreply.github.com'
 
 jobs:
+  add-version-to-docs:
+    if: startsWith(github.ref, 'refs/tags/')
+    name: 'Handle `versions.json` in `gh-pages` branch'
+    runs-on: ubuntu-20.04
+    steps:
+      - name: 'Checkout the default branch (master)'
+        uses: actions/checkout@v2
+
+      - name: 'Checkout the docs branch to a subdirectory'
+        uses: actions/checkout@v2
+        with:
+          ref: gh-pages
+          path: gh-pages
+
+      # Updates `env.DOCS_VERSION` to the tag version; if invalid exits job early.
+      - name: 'Ensure `versions.json` has relevant `<major>.<minor>` tag version (truncates patch version)'
+        id: add-version
+        continue-on-error: true
+        working-directory: gh-pages
+        run: '../.github/workflows/scripts/docs/update-versions-json.sh'
+
+      # If an actual change was made to `versions.json`, commit and push it - otherwise the step is skipped instead of reporting job failure.
+      - name: 'Push update for `versions.json`'
+        if: ${{ steps.add-version.outcome == 'success' }}
+        working-directory: gh-pages
+        run: |
+          git config user.name ${{ env.GIT_USER }}
+          git config user.email ${{ env.GIT_EMAIL }}
+          git add versions.json
+          git commit -m "Make ${{ env.DOCS_VERSION }} available to the 'mike' version selector"
```

Comment:
Maybe use conventional commits syntax?
`git commit -m "docs(version): add ${{ env.DOCS_VERSION }} to selector"`